Zend/zend_alloc.c

```diff
@@ -296,6 +296,20 @@ static bool zend_mm_use_huge_pages = false;
  */
 
 struct _zend_mm_heap {
+    union {                                             /* This union contains security-relevant properties, and is thus made read-only at run-time. */
+        struct {
+            uintptr_t          shadow_key;              /* free slot shadow ptr xor key */
+#if ZEND_MM_CUSTOM
+            struct {
+                void      *(*_malloc)(size_t ZEND_FILE_LINE_DC ZEND_FILE_LINE_ORIG_DC);
+                void       (*_free)(void*  ZEND_FILE_LINE_DC ZEND_FILE_LINE_ORIG_DC);
+                void      *(*_realloc)(void*, size_t  ZEND_FILE_LINE_DC ZEND_FILE_LINE_ORIG_DC);
+            } custom_heap;
+#endif
+        };
+        char padding [ZEND_MM_PAGE_SIZE];
```

Comment:
The heap is embedded in `zend_mm_segment`. Currently we rely on the fact that `zend_mm_segment` is smaller than `ZEND_MM_PAGE_SIZE * ZEND_MM_FIRST_PAGE`. If we significantly increase the size of `zend_mm_heap` we need to either increase `ZEND_MM_FIRST_PAGE`, or to move it out of `zend_mm_segment`. The latter is probably better, otherwise we lose `ZEND_MM_PAGE_SIZE * ZEND_MM_FIRST_PAGE` bytes in every segment.

(I've been hit by this while experimenting. We should add a `ZEND_STATIC_ASSERT(sizeof(zend_mm_segment) <= ZEND_MM_PAGE_SIZE * ZEND_MM_FIRST_PAGE)` somewhere.)

Reply:
What's `zend_mm_segment`?

Reply:
Oh I meant zend_mm_chunk, sorry